clang/lib/AST/TypePrinter.cpp

```diff
@@ -852,34 +852,50 @@ void TypePrinter::printExtVectorAfter(const ExtVectorType *T, raw_ostream &OS) {
 
 void TypePrinter::printConstantMatrixBefore(const ConstantMatrixType *T,
                                             raw_ostream &OS) {
+  if (Policy.UseHLSLTypes)
+    OS << "matrix<";
   printBefore(T->getElementType(), OS);
-  OS << " __attribute__((matrix_type(";
-  OS << T->getNumRows() << ", " << T->getNumColumns();
-  OS << ")))";
```

Comment:
This change seems wrong to me. IIUC this changes something like `float  __attribute__((matrix_type(4, 4)))*` to `float * __attribute__((matrix_type(4, 4)))`, which would mean the element type is `float*` rather than `float`.

Reply:
See my response to Florian [here](https://github.com/llvm/llvm-project/pull/111415/files#r1796090424)

Reply:
may be simpler to read to duplicate some code but have the HSL and C++ matrixes types printed completely separately?